Bluetooth: host: add missing meta data destroy callback

jori-nordic opened the pull request:
Symptoms were that a disconnect happening when a buffer was enqueued but
not sent resulted in a meta-data memory leak.

This is problematic because it seemed (as per the OP) that it resulted in
the whole host getting locked up when a device sending long L2CAP packets
was prone to disconnect-reconnect cycles (e.g. in a bad RF environment).

Fixes #47649

Signed-off-by: Jonathan Rico <jonathan.rico@nordicsemi.no>

jori-nordic committed:
Bluetooth: host: add missing meta data destroy callback

Symptoms were that a disconnect happening when a buffer was enqueued but
not sent resulted in a meta-data memory leak.

This is problematic because it seemed (as per the OP) that it resulted in
the whole host getting locked up when a device sending long L2CAP packets
was prone to disconnect-reconnect cycles (e.g. in a bad RF environment).

Fixes #47649

Signed-off-by: Jonathan Rico <jonathan.rico@nordicsemi.no>

carlst commented:
@jori-nordic I can no longer reproduce the issue with https://github.com/zephyrproject-rtos/zephyr/pull/48395 applied. And I verified the patch code was called (and the in-flight buffer freed by ATT). A nice, clean fix.

jori-nordic commented:
@Thalley could you check the Bluetooth Tests CI step ? I'm not sure the failure is due to this patch.

Thalley commented:
> @Thalley could you check the Bluetooth Tests CI step ? I'm not sure the failure is due to this patch.

Hmm, it's a segmentation fault caused when sending, so it may very well be related to this PR. 

Not sure if this PR causes the issue, or simply exposes the issue. 

I will suggest to use valgrind with the babblesim test to help find the issue

sjanc commented on subsys/bluetooth/host/conn.c:
```diff
@@ -768,6 +768,10 @@ void bt_conn_process_tx(struct bt_conn *conn)
 	BT_ASSERT(buf);
 	if (!send_buf(conn, buf)) {
+		if (tx_data(buf)->tx) {
+			conn_tx_destroy(conn, tx_data(buf)->tx);
```
After looking a bit more at the code I think you should also set tx to NULL after destroying it.
There are two more conditional calls to conn_tx_destroy() and those (conn_cleanup and send_frag) probably also should set tx to NULL...

jori-nordic replied:
I looked at it too, but why should I do that ? is it so nothing uses the (destroyed) tx reference in the rest of the function (s) ? because the buffer will get destroyed and along with it its user data (where tx is stored).

jori-nordic replied:
nvm, you're right: the failing test was because of this succession of events:
- `send_frag` failed, then called `conn_tx_destroy` but still kept the `tx` ref, because `always_consume` wasn't set.
- `bt_conn_process_tx` also failed, and attempted to destroy the (now invalid) `tx` ref -> segfault.

alwa-nordic replied:
The part about the callbacks in `conn_tx_destroy` still stands. It may be safer to detach `tx` first. We have adopted this pattern several places already.

```c
struct bt_conn_tx tx = tx_data(buf)->tx;
tx_data(buf)->tx = NULL;
conn_tx_destroy(conn, tx);
```

This also allows moving the `net_buf_unref` up. Freeing earlier is always better. (It's like releasing a lock.)
```c
struct bt_conn_tx detached_tx = NULL;
if (buf->ref == 1) {
	detached_tx = tx_data(buf)->tx;
	/* The following is not strictly needed because net_buf_unref invalidates `tx_data(buf)->tx`. */
	tx_data(buf)->tx = NULL;
}
net_buf_unref(buf);
if (detached_tx) {
	conn_tx_destroy(conn, tx);
}
```

sjanc commented:
yeap, I was able to reproduce this locally
```
[janc@ix zephyr]$ git diff
diff --git a/subsys/bluetooth/host/conn.c b/subsys/bluetooth/host/conn.c
index 6600dae5c7..89b3ce7071 100644
--- a/subsys/bluetooth/host/conn.c
+++ b/subsys/bluetooth/host/conn.c
@@ -562,6 +562,7 @@ fail:
        k_sem_give(bt_conn_get_pkts(conn));
        if (tx) {
                conn_tx_destroy(conn, tx);
+               tx_data(buf)->tx = NULL;
        }
 
        if (always_consume) {
@@ -675,6 +676,7 @@ static void conn_cleanup(struct bt_conn *conn)
        while ((buf = net_buf_get(&conn->tx_queue, K_NO_WAIT))) {
                if (tx_data(buf)->tx) {
                        conn_tx_destroy(conn, tx_data(buf)->tx);
+                       tx_data(buf)->tx = NULL;
                }
 
                net_buf_unref(buf);
@@ -768,6 +770,10 @@ void bt_conn_process_tx(struct bt_conn *conn)
        buf = net_buf_get(&conn->tx_queue, K_NO_WAIT);
        BT_ASSERT(buf);
        if (!send_buf(conn, buf)) {
+               if (tx_data(buf)->tx) {
+                       conn_tx_destroy(conn, tx_data(buf)->tx);
+                       tx_data(buf)->tx = NULL;
+               }
                net_buf_unref(buf);
        }
 }
```

jori-nordic commented:
@Thalley @sjanc I'll look into it then, thanks

jori-nordic committed:
Bluetooth: host: invalidate tx context after destroying it

Now that we properly destroy the tx context in `bt_conn_process_tx`, we
need to make sure the context is really valid.

Signed-off-by: Jonathan Rico <jonathan.rico@nordicsemi.no>

hermabe commented on subsys/bluetooth/host/conn.c:
```diff
@@ -562,6 +562,10 @@ static bool send_frag(struct bt_conn *conn, struct net_buf *buf, uint8_t flags,
 	k_sem_give(bt_conn_get_pkts(conn));
 	if (tx) {
 		conn_tx_destroy(conn, tx);
+		/* The buffer might not get destroyed, make sure that this tx context is not
+		 * re-used later.
+		 */
```
Is this comment accurate? `conn_tx_destroy()` does not free the buffer, only the metadata. The buffer itself is freed by `net_buf_unref()`. As I understand it, we want to set `tx = NULL` so that `conn_tx_destroy()` is not called twice for the same `tx`.

hermabe replied:
Then I think we agree about the reasoning here. Your updated comment is better.

alwa-nordic commented on subsys/bluetooth/host/conn.c:
```diff
@@ -768,6 +773,12 @@ void bt_conn_process_tx(struct bt_conn *conn)
 	BT_ASSERT(buf);
 	if (!send_buf(conn, buf)) {
+		/* buf is going to be destroyed by the next unref */
+		if (buf->ref == 1 && tx_data(buf)->tx) {
```
I'm not sure about the `buf->ref == 1` requirement. I think the TX callbacks must be invoked by whatever process removed `buf` from `tx_queue`. In this code branch, `buf` is not put back on `tx_queue`. Am I wrong?

hermabe replied:
The TX thread does not care if `send_buf()` succeeds or not and does not put the buffer back in the `tx_queue` if it fails.

jori-nordic replied:
yeah its not put back, I'll remove it.

sjanc commented on subsys/bluetooth/host/conn.c:
```diff
@@ -768,7 +773,17 @@ void bt_conn_process_tx(struct bt_conn *conn)
+
+		/* destroy the tx context (and any associated meta-data) */
+		if (tx) {
+			conn_tx_destroy(conn, tx_data(buf)->tx);
```
shouldn't this now be conn_tx_destroy(conn, tx); ?

Thalley commented on subsys/bluetooth/host/conn.c:
```diff
@@ -562,6 +562,10 @@ static bool send_frag(struct bt_conn *conn, struct net_buf *buf, uint8_t flags,
+		/* `buf` might not get destroyed, and its `tx` pointer will still be reachable.
+		 * Make sure that we don't try to use the destroyed context later.
+		 */
+		tx_data(buf)->tx = NULL;
```
This PR kind of does this part in 3 different ways. 

The first one here we store `tx_data(buf)->tx` as a local variable and use that for the NULL check and the call for `conn_tx_destroy`, and then set `tx_data(buf)->tx = NULL;`. 

The next one is arguable simpler and does 
```c
		if (tx_data(buf)->tx) {
			conn_tx_destroy(conn, tx_data(buf)->tx);
			tx_data(buf)->tx = NULL;
		}
```
and is very easy to follow. 

Then you have the third one where you first store it as a local variable `struct bt_conn_tx *tx = tx_data(buf)->tx;`, then set `tx_data(buf)->tx = NULL;`, then do the NULL check on the local variable and then finally call `conn_tx_destroy(conn, tx);`

Might be worth to follow the same pattern in all 3 for the sake of readability.

jori-nordic replied:
very much agree. I originally had (and preferred) number 2 (the one with the code block) but @alwa-nordic wanted me to use num 3 instead.

I would've preferred the second one because it's simpler to read, but `send_frag` reads better with the `tx` local, so I'l use the third method everywhere.

jori-nordic commented:
@Thalley @sjanc could you re-review?
